### PR TITLE
Handle unaligned FP8 block quantization edge case

### PR DESCRIPTION
**Pull request by catswe**
# Description

Handled edge case in linked issue https://github.com/vllm-project/tpu-inference/issues/1121#issuecomment-3706449354.

# Tests

Verified correctness manually by testing against `gaunernst/DeepSeek-V2-Lite-Chat-FP8`, a test model for `DeepSeek-V3/R1` (getting this model to work required additional changes not included in this PR, see https://github.com/vllm-project/tpu-inference/pull/1593). Also verified that the existing code fails under the new test case, `test_unaligned_block_quantization`, while the current changes passes.

```
JAX_TRACEBACK_FILTERING=off VLLM_MLA_DISABLE=1 VLLM_LOGGING_LEVEL=DEBUG MODEL_IMPL_TYPE=vllm vllm serve gaunernst/DeepSeek-V2-Lite-Chat-FP8   --max-model-len=4096   --gpu-memory-utilization=0.98  --download-dir /dev/shm
```

Screenshot 2026-01-31 at 10 34 29 PM

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/common/quantization/__init__.py [hunk 1]
```diff
@@ -165,6 +165,39 @@ def dequantize_tensor_from_mxfp4_packed(
     )
 
 
+def pad_and_dequantize_tensor(
```
Comment (kyuyeunk) [inline tpu_inference/layers/common/quantization/__init__.py]: awesome generic solution that is applicable for variety of situations!

File: tests/layers/vllm/test_fp8.py [hunk 2]
```diff
@@ -535,3 +552,37 @@ def test_blockwise_quant(requant_block_size, requant_weight_dtype):
...
+            output_size=output_size,
+            bias=False,
+            params_dtype=dtype,
+            return_bias=False,
+            quant_config=quant_config,
+        )
+
+    initialize_layer_weights(linear_layer)
+    ref_output, layer_output = return_ref_and_layer_output(linear_layer)
+    torch.testing.assert_close(ref_output, layer_output)
```
Comment (kyuyeunk) [inline tests/layers/vllm/test_fp8.py]: add new line

File: tpu_inference/layers/common/quantization/__init__.py (hunk 1 above)
Comment (kyuyeunk) [inline tpu_inference/layers/common/quantization/__init__.py]: how about we move this functionality into the main function  - `dequantize_tensor` - so that padding / slicing is applied automatically if alignment issue is detected? (i.e., dim size is not divisible by block size?) but let me know if it makes the main function too complex / spaghetti code.

Comment (kyuyeunk) [inline tpu_inference/layers/common/quantization/__init__.py]: please add a test for this function in https://github.com/vllm-project/tpu-inference/blob/main/tests/layers/common/test_quantization.py

Comment (catswe) [reply tpu_inference/layers/common/quantization/__init__.py]: Thank you!

File: tpu_inference/layers/common/quantization/__init__.py [hunk 3]
```diff
@@ -106,18 +108,26 @@ def dequantize_tensor(
...
+
+    aligned_shape = tensor_q.shape
     if tensor_q.ndim == scale.ndim:
         # Indicates the tensor was block quantized.
-        blocked_shape = [[i] for i in orig_shape]
+        blocked_shape = [[i] for i in aligned_shape]
         for i in axis:
             num_blocks = scale.shape[i]
             if tensor_q.shape[i] % num_blocks:
                 raise ValueError(
```
Comment (kyuyeunk) [inline tpu_inference/layers/common/quantization/__init__.py:125]: just to double check: if `block_size` is not passed as an argument and the tensor shape does not align with `num_blocks` found in `scale.shape[i]`, it will throw an error right? is there an unit test that checks if error is thrown in this scenario?

Comment (catswe) [reply tpu_inference/layers/common/quantization/__init__.py:125]: There isn't any unit test for this. I commented out the `raise ValueError`, and all of them still passed, so I added another test. This path isn't actually very robust, though. If `tensor_q` has shape `(100, 100)` and scale has shape `(4, 4)`, the error will never get raised since 100 is perfectly divisible by 4. Let me know if you want to do something like

```python
if calc_block_size not in [16, 32, 64, 128, 256, 512]:
    raise ValueError("Block size must be a valid power of 2 from 16 to 512")
```

Comment (kyuyeunk) [reply tpu_inference/layers/common/quantization/__init__.py:125]: also. can you slightly tweak the error message that basically says: "if you want to perform unaligned dequantization, please pass block size as an argument."

Comment (kyuyeunk) [reply tpu_inference/layers/common/quantization/__init__.py:125]: >  If tensor_q has shape (100, 100) and scale has shape (4, 4), the error will never get raised since 100 is perfectly divisible by 4. Let me know if you want to do something like

I think that's okay! the default expectation is that tensor shapes should always aligned. so for the example you've described, i think we can just say it's an as: "intended behavior. if you want a proper dequantization with unaligned case, make sure to pass in block size as well"

**Review by kyuyeunk — APPROVED** [review]

**Comment by kyuyeunk** [comment]
thank you for the awesome change! triggered ci. please ping me if i forget to press merge even when ci has been passed.

**Comment by catswe** [comment]
I removed the "simulate varying scales" code block since it was just a nice extra check. It causes repeated failing fp8 quant tests on TPU7x (seems like TPU7x is a lot more sensitive numerically since the same one without TPU7x is fine?). I don't really want to adjust test tolerances away from default just to account for this either. In practice, new bugs would likely still be caught by other errors raised by the test

**Comment by kyuyeunk** [comment]
> I removed the "simulate varying scales" code block since it was just a nice extra check. It causes repeated failing fp8 quant tests on TPU7x (seems like TPU7x is a lot more sensitive numerically since the same one without TPU7x is fine?). I don't really want to adjust test tolerances away from default just to account for this either. In practice, new bugs would likely still be caught by other errors raised by the test

got it. that makes sense to me. anyway, merged!
